OMCompiler/SimulationRuntime/c/util/doubleEndedList.c

```diff
@@ -0,0 +1,534 @@
+ * ACCORDING TO RECIPIENTS CHOICE.
+ *
+ * The OpenModelica software and the OSMC (Open Source Modelica Consortium)
+ * Public License (OSMC-PL) are obtained from OSMC, either from the above
+ * address, from the URLs: http://www.openmodelica.org or
+ * http://www.ida.liu.se/projects/OpenModelica, and in the OpenModelica
+ * distribution. GNU version 3 is obtained from:
+ * http://www.gnu.org/copyleft/gpl.html. The New BSD License is obtained from:
+ * http://www.opensource.org/licenses/BSD-3-Clause.
+ *
+ * This program is distributed WITHOUT ANY WARRANTY; without even the implied
+ * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE, EXCEPT AS
+ * EXPRESSLY SET FORTH IN THE BY RECIPIENT SELECTED SUBSIDIARY LICENSE
+ * CONDITIONS OF OSMC-PL.
+ *
+ */
+
+/*! \file doubleEndedList.c
+ *
+ * Description: This file contains a simple double ended linked list.
+ */
+
+#include "doubleEndedList.h"
+#include "omc_error.h"
+
+#include <stdlib.h>
+#include <string.h>
+
+
+/**
+ * @brief A single node element of a double ended list.
+ *
+ * Knows previous and next node and has some data.
+ */
+struct DOUBLE_ENDED_LIST_NODE {
+  void* data;                      /**< Item data */
+  DOUBLE_ENDED_LIST_NODE* prev;    /**< Pointer to previous node in list */
+  DOUBLE_ENDED_LIST_NODE* next;    /**< Pointer to next node in list */
+};
+
+
+/**
+ * @brief Double ended list.
+ *
+ * Has pointers to first and last element and can be iterated over forward and backward.
+ *
+ */
+struct DOUBLE_ENDED_LIST {
+  DOUBLE_ENDED_LIST_NODE* first;   /**< Pointer to first element of list */
+  DOUBLE_ENDED_LIST_NODE* last;    /**< Pointer to last element of list */
+  unsigned int itemSize;           /**< Size of item data */
+  unsigned int length;             /**< Number of elements in list */
+};
+
+
+// ############################################################################
+//
+// Section for allocating and freeing double ended list
+//
+// ############################################################################
+
+
+/**
+ * @brief Create new empty double ended list.
+ *
+ * @param itemSize              Size of item data.
+ * @return DOUBLE_ENDED_LIST*   Pointer to new created double ended list.
+ */
+DOUBLE_ENDED_LIST* allocDoubleEndedList(unsigned int itemSize) {
+  DOUBLE_ENDED_LIST* list = (DOUBLE_ENDED_LIST*) malloc(sizeof(DOUBLE_ENDED_LIST));
+  list->first = NULL;
+  list->last = NULL;
+  list->itemSize = itemSize;
+  list->length = 0;
+
+  return list;
+}
+
+
+/**
+ * @brief Free double ended list.
+ *
+ * Frees list items as well.
+ *
+ * @param list    Pointer to list to be freed.
+ */
+void freeDoubleEndedList(DOUBLE_ENDED_LIST *list) {
+  if(list) {
+    clearDoubleEndedList(list);
+    free(list);
+  }
+}
+
+
+/**
+ * @brief Free double ended list node.
+ *
+ * @param node    Pointer to list node.
+ */
+void freeNodeDoubleEndedList(DOUBLE_ENDED_LIST_NODE *node) {
+  free(node->data);
```

Comment:
This may cause leaks if `data` contains allocated memory. You could add a `destroy_call(node->data)` before `free`ing or something.

I would leave it like this but add a note of caution somewhere. Who knows what people will want to do with this :slightly_smiling_face: